Implementation of lossy cast lints (issue #114)

Robzz opened the pull request:
This PR introduces the 3 lints suggested by @llogiq in #114 along with a test case.

There must be a way to refactor the code (some match arms are exactly the same except for a variant), but I'm no expert in matching patterns and I haven't found how. I thought of macros, but it seems way overkill. Advice appreciated.

With dogfood.sh, the cast_precision_loss lint triggered 2 errors in src/consts.rs. I silenced them because although there is a loss of precision, the code seems legit.

Also, keep in mind that as I matched on the actual type variants (I didn't find a way to ask rustc for the type's size), this is not generic and will need updating as larger types are added to rust.

And there goes my first nontrivial contribution to an open-source project :smile:

birkenfeld commented on src/types.rs:
```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+                    (false, true) => {
+                        span_lint(cx, CAST_POSSIBLE_OVERFLOW, expr.span, 
+                                  &format!("the contents of a {} may overflow a {}", cast_from, cast_to));
+                        if !cx.tcx.expr_ty(expr).is_signed() {
```
this is `cast_to.is_signed()`?

Robzz replied:
Yes, good catch!

birkenfeld commented on src/types.rs:
```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+    fn check_expr(&mut self, cx: &Context, expr: &Expr) {
+        if let ExprCast(ref ex, _) = expr.node {
+            let (cast_from, cast_to) = (cx.tcx.expr_ty(&*ex), cx.tcx.expr_ty(expr));
+            if cast_from.is_numeric() && !in_external_macro(cx, expr.span) {
```
you're only checking `is_numeric` for `cast_from`. I assume that any casts to non-int/float are rejected by the compiler first?

Related question: is there a trait you can `impl` for casts? Might there be in the future?

Manishearth replied:
lints run post-typeck (post-everything really, I think the only further errors can come from translation and/or LLVM)

But it is totally possible to do something like `1u8 as *const u8`, so we should be checking both arms.

`CoerceUnsized` exists, which lets you widen the range of coercions, and I think this indirectly widens the range of casts too, but I'm not sure.

birkenfeld commented on src/types.rs:
```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+                        match (&cast_from.sty, &cast_to.sty) {
+                            (&ty::TypeVariants::TyInt(i1), &ty::TypeVariants::TyInt(i2)) => {
+                                match (i1, i2) {
+                                    (ast::IntTy::TyI64, ast::IntTy::TyI32) |
```
What do we do about `TyIs` and `TyUs`? We could treat it like either 32 or 64 bits and warn when either width would cause problems.

I.e. `*64 as *size`, `*size as *32` or smaller, `usize as isize`.

llogiq replied:
Perhaps treat *size as 64 bits by default and add another lint that matches when compiling on 32 bits would overflow?

birkenfeld commented:
You have a few lines with trailing whitespace. Otherwise LGTM.

llogiq commented:
Thank you for your contribution! As far as I can see the code looks good so far.  The `#[allow(..)]`s in `const.rs` are to be expected.

The duplication could be reduced by first destructuring the types and casting the `TyI*`/`TyU*` to usize (this will cast `Ty*s` to `0`, Ty*8` to `1`, etc.). The number of bits is then (apart from the `*size` case, which should be handled separately anyway) `4 << ty as usize`.

Apart from the whitespace nit, r+, unless you want to address those other things first.

Manishearth commented on src/consts.rs:
```diff
@@ -67,15 +67,16 @@ impl Constant {
-           _ => None
-       }
-   }
+    #[allow(unknown_lints,cast_precision_loss)]
```
nit: `#![allow(unknown_lints)]` should be at the crate level

Robzz replied:
Fixed, thanks!

Manishearth commented on src/types.rs:
```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+                                    (ast::IntTy::TyI64, ast::FloatTy::TyF32) |
+                                    (ast::IntTy::TyI64, ast::FloatTy::TyF64) => {
+                                        span_lint(cx, CAST_PRECISION_LOSS, expr.span,
+                                                  &format!("converting from {} to {}, which causes a loss of precision",
```
nit: perhaps some more text here explaining why this is a loss of precision? (intX has X bits for the mantissa, floatX has less than X bits for the mantissa)

llogiq replied:
Also we may want to add a link to our wiki with a deeper explanation + perhaps some examples.

llogiq commented on src/types.rs:
```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+declare_lint!(pub CAST_SIGN_LOSS, Allow,
+              "casts from signed types to unsigned types");
+declare_lint!(pub CAST_POSSIBLE_OVERFLOW, Allow,
+              "casts that may cause overflow");
```
We may want to add examples here.

```
"casts that cause loss of precision, e.g `x as f32` where `x: u64`"
"casts from signed types to unsigned types, e.g. `x as u32` where `x: i16`"
"casts that may cause overflow, e.g. `x as u8` where `x: u32`"`
```

Manishearth replied:
+1, I was suggesting the same thing for the outputted error message. These errors can be subtle, esp the first one.

Robzz replied:
That's a good idea, I added it :)

Manishearth commented on src/types.rs:
```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+                            }
+                            (&ty::TypeVariants::TyUint(u), &ty::TypeVariants::TyFloat(f)) => {
+                                match (u, f) {
+                                    (ast::UintTy::TyU32, ast::FloatTy::TyF32) |
```
nit: import `ast::*` or just `ast::UintTy`/FloatTy here.

Also, I think the AST reexports `TyU32` and all, so `ast::TyU32` should just work.

Robzz replied:
You are right, I didn't even think of looking, I fixed that now. Will push soon.
As s side note, middle::ty also reexports TypeVariants::*, so I also turned &ty::TypeVariants::Ty* into &ty::Ty*

Manishearth commented:
Thanks for the PR, and I'm happy that Clippy could help you contribute more to open source! (as a first contribution, messing with compiler internals is pretty awesome).

Like the others have said, couple of nits (and one small bug), and we should be good.

Robzz commented:
Okay I think I addressed most of your remarks. What remains is :

1. some more helpful messages
2. reduce duplication (thanks for the tip @llogiq )
3. the TyS/TyUs case, but I'm not really sure how to handle that.

I will be back in a few hours to address 1 and 2, what to do for 3 is up to you guys :smile:

llogiq commented:
:+1: for merge without `*size` handling once the other things are finished.

We can add the arch-dependent lint variant later.

llogiq commented on src/types.rs:
```diff
@@ -136,3 +136,136 @@ impl LintPass for UnitCmp {
+declare_lint!(pub CAST_SIGN_LOSS, Allow,
+              "casts from signed types to unsigned types, e.g `x as u32` where `x: i32`");
+declare_lint!(pub CAST_POSSIBLE_OVERFLOW, Allow,
+              "casts that may cause overflow, e.g `x as u8` where `x: u32`, or `x as i32` where `x: f32`");
```
Thinking a bit more about this, the cast will *not* cause overflow. It will just silently truncate the value. Perhaps we should rename this to `CAST_TRUNCATE` and explain `"casts that truncate the value, ..."`?

Robzz commented:
It should be good this time. Summary : 

* I refactored the whole thing to reduce code duplication, and shrinked the file by no less than 50 lines. Great tip @llogiq :smiley: 
* I added some more information to the precision loss lint as requested by @Manishearth. I think the "mathy" way will do, as the lint is opt-in and users should hopefully have an idea of what they are looking for if they enable it, but as you mentioned, a more detailed explanation on the wiki could not hurt.
* I took @llogiq's remarks about the cast_possible_overflow lint into account and renamed it to cast_possible truncation, and updated the message as well.
* Updated the test case because of the new messages.

Let me know if I forgot something!

birkenfeld commented on src/types.rs:
```diff
@@ -136,3 +136,86 @@ impl LintPass for UnitCmp {
+                match (cast_from.is_integral(), cast_to.is_integral()) {
+                    (true, false) => {
+                        let from_nbits = match &cast_from.sty {
+                            &ty::TyInt(i) =>  4 << (i as usize),
```
This looks correct, but deserves a comment IMO.

Robzz replied:
Correct, I thought of it after the fact. The default arm of the match also somewhat bothers me, because although it should never be reached (is_integral is only true for ty::IntTy and ty::UintTy according to rustc source), if it ever was, then it would successfully pass the `if from_nbits != 4` and cause shenanigans.

Like I said, it shouldn't happen but I'm considering putting 4 as the default match value to prevent it, or check and ensure that `from_nbits != 0`. Anyway, I will document it as what happens there is nowhere near obvious.

birkenfeld replied:
> The default arm of the match also somewhat bothers me, because although it should never be reached

That could use a `unreachable!()` call.

birkenfeld commented on src/types.rs:
```diff
@@ -136,3 +136,86 @@ impl LintPass for UnitCmp {
+                        }
+                    },
+                    (false, true) => {
+                        span_lint(cx, CAST_POSSIBLE_TRUNCATION, expr.span,
```
Good question: how exactly is the floating point value "truncated" when cast to integral types? Apparently `1e100f64 as u64` is just zero...

Robzz replied:
Good question indeed, it is answered in the [rustonomicon](https://doc.rust-lang.org/nightly/nomicon/casts.html). According to this, your example (currently) invokes undefined behaviour as the float contents do not fit an int.
Note that I just stumbled onto that as I was looking for an answer to the question. I'll need to read it in more detail to see if the lints are correct, or if I can improve the messages some more.

birkenfeld replied:
Ah, nice!

birkenfeld commented on src/types.rs:
```diff
@@ -136,3 +136,86 @@ impl LintPass for UnitCmp {
+                            span_lint(cx, CAST_SIGN_LOSS, expr.span,
+                                      &format!("casting from {} to {} loses the sign of the value", cast_from, cast_to));
+                        }
+                        let from_nbits = match &cast_from.sty {
```
this `match` appears 3 times: it should be put into a utility function (then it can also get the appropriate comments about why it works :)

Robzz replied:
Good idea, will do. Thanks!

Robzz replied:
Bonus : this is a very good place to handle the case of isize/usize. I'm doing that right now, and it seems I will be able to include it in this PR.

llogiq commented:
@Robzz Here's hoping I never get to see, nay, browse such a string literal. :smile:

I'm ok with the "mathy" explanation, but we should add something for non-math-affine people on the wiki. Once I have the basic structure done, this will be part of #210.

The only thing that I'm a bit unhappy about is the reliance on target system pointer width – I think we should not assume that the target pointer width will always be the same as used by the current system. In `const.rs`, we refuse to guess, too. I think we should treat *x*`size` as a *superposition* of *x*`32`, *x*`64` and report all errors that *may* arise, if possible also noting the pointer sizes under which the problem arises (if not all).

However, this can be done in a separate PR, so: Well done :+1:, r+

llogiq commented:
@Robzz It may interest you that I've added a few words on the lints on the wiki:

* [`cast_possible_truncation`](https://github.com/Manishearth/rust-clippy/wiki#cast_possible_truncation)
* [`cast_precision_loss`](https://github.com/Manishearth/rust-clippy/wiki#cast_precision_loss)
* [`cast_sign_loss`](https://github.com/Manishearth/rust-clippy/wiki#cast_sign_loss)

Robzz commented:
About the pointer width, I didn't look at it that way but you're right. If I understand correctly, you're saying that it would make more sense warning when the cast may behave differently on different archs than the current behaviour, which is to warn when the cast may be problematic on the current target arch. If so, it makes sense to me. Heh, my mistake, so I'll fix it, I'll send a PR in the next few days, as I'd rather work on something else for a little while. To be sure, would you rather have it in the current lint or in another one (such as, say, `cast_arch_dependant`)? This seems a bit more severe to me than a regular possible truncation, maybe we should even `Warn` by default. Your call.

Working on this made me learn a bit about the compiler internals and I'm looking forward to learn more, so I guess I'll find myself another issue to chew on. Anyway, big thanks to the three of you for your niceness and helpfulness! :smile:

(Gosh, I'm realizing this PR discussion is getting fairly long, I'll try being a bit more efficient next time :) )

llogiq commented:
I'd just add it to the existing lint, but change the message to let the user know whether the *problem* is arch dependent, and if so on which arches it appears. My reasoning is that the user has knowingly casted to a `*size`, so they already know the target width is arch-dependent, while truncation will usually not be a problem here.

> Heh, my mistake, so I'll fix it...

I don't see this as a mistake, but as a possible extension. Feel free to PR whenever you like.

> Anyway, big thanks to the three of you for your niceness and helpfulness! :smile:

Thank *you* for your work on clippy! :smile:

> I'll try being a bit more efficient next time

It's alright. No need to strive for brevity at the expense of good communication.

birkenfeld commented:
Thanks for the PR!
